### ISSUE TITLE
Warn when promises are used as value

### ISSUE THREAD

**Issue by steelbrain** [facebook/flow#3862]
Flow should warn when we try to use a promise as a value, instead of `.then` or `await`ing on them. Some scenarios where the warning should be shown:

```js
if (promiseReturningFunction()) {}
```
```js
const something = promiseReturningFunction()
something.greet()
```

The scenarios below however should still be okay

```js
if (await promiseReturningFunction()) {}
```
```js
const something = promiseReturningFunction()
const anotherThing = await something
anotherThing.greet()
```

Let me know if I any further explanation is required. Thank you all for working on Flow!

**Comment by vkurchatkin** [comment]
What about just:

```js
promiseReturningFunction()
```
this should be an error too?

**Comment by steelbrain** [comment]
It should be. Because you are not handling the rejection at all. An acceptable way to write it would be this

```js
promiseReturningFunction().catch(e => console.error(e))
```

**Comment by vkurchatkin** [comment]
What if rejection is impossible and you know it? Anyway, you can't attach `catch` to all promises you create for obvious reasons

**Comment by steelbrain** [comment]
> Anyway, you can't attach `catch` to all promises you create for obvious reasons

You can't, but you don't dispatch a promise and not care about it's results for all promises for obvious reasons as well. In the extremely rare cases when there's zero possible points of failure and you know it, you can use a suppress comment. Using a suppress comment in a few places is better than having flow not catch errors in the rest of the codebase.

**Comment by steelbrain** [comment]
> Flow catches type errors with promises just fine. What you propose is just too opinionated.

You are right. I have trimmed down the initial post to something that I think we can all agree on. Let me know what you think

**Comment by vkurchatkin** [comment]
Well this case should be covered by reachability analysis, because if `promiseReturningFunction()` is known to always return a promise, then `if` is excess

**Comment by steelbrain** [comment]
It should be. But flow doesn't report it at the time of writing.

<img src="https://cloud.githubusercontent.com/assets/4278113/25641146/04962152-2fac-11e7-884c-91667c45d05e.png">

**Comment by vkurchatkin** [comment]
Right, but there are open issues for this

**Comment by steelbrain** [comment]
My search fu is failing because of the massive amount of open issues and not knowing the exact keyword to look for, could you please link me to some of them? I'd happily close in their favor
